### ISSUE TITLE
Segmentation fault after repeated simulations

### ISSUE THREAD

**Issue by tasseff** [USEPA/WNTR#115]
I'm currently using WNTR within an optimization routine that iteratively solves EPANET models. I'm finding that, consistently, after 1016 iterations, WNTR (or one of its libraries) is producing a segmentation fault. This is quite mysterious, as this behavior persists even when simulating over a completely different network. Here's [a link to one such INP file that I'm using](http://emps.exeter.ac.uk/media/universityofexeter/emps/research/cws/downloads/hanoi.inp). A simple script that should produce this error is found below. Notably, this is happening on Ubuntu 18.04 LTS, and WNTR was installed using `pip install`.

```python
import wntr

network = wntr.network.WaterNetworkModel('hanoi.inp')

for i in range(0, 2000):
    sim = wntr.sim.EpanetSimulator(network)
    results = sim.run_sim()
    print(i)
```

**Comment by dbhart** [comment]
I attempted the same thing using the code you provided and when using Net3 from EPANET. (BTW, there are negative pressures at timestep 0 in your model). I reproduced the results you are seeing with both. I turned on debugging, and the following code shows the results.
The EPANET error 305 indicates that it was unable to open the file. Because logging prevented it from doing a segfault, I was able to try doing the loop again immediately afterwards, and got an OSError - too many files open. This is a known issue with EPANET 2.0.12 - there are files that do not get closed properly; I thought I had managed to fix it in the .so file distributed with WNTR, but clearly I haven't. We are in the process of integrating EPANET 2.2, which is supposed to have fixed this issue, but are not there yet.

There may be a toolkit function that we can add to try to get the files closed, but we've found that we created segfaults in a different place when we do that, so getting the EPANET 2.2 library in will be the best solution. There are some other (ugly) solutions you can try in the meantime:

- use the wntr.epanet.io functions to write the INP file and read the BIN file, but call a system command to run the runepanet command line program in between (this will also allow you to use EPANET 2.2 before we get it integrated - it is backwards compatible with 2.0 INP files).
- use a subprocess to call the sim.run_sim() in a separate python process, and send back the results. I have used this method personally before, but it is a lot of work.

We will definitely look into this.

```
>>> import wntr
>>> wntr.start_logging()
>>> network = wntr.network.WaterNetworkModel('Net3.inp')
>>> for i in range(0, 2000):
...     sim = wntr.sim.EpanetSimulator(network)
...     results = sim.run_sim(save_hyd=True)
...     print(i)
... 
0
1
2
###### extra lines cut out 
241
242
243
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
  File "/Users/dbhart/Projects/git-wntr/wntr/sim/epanet.py", line 89, in run_sim
  File "/Users/dbhart/Projects/git-wntr/wntr/epanet/toolkit.py", line 290, in ENsavehydfile
  File "/Users/dbhart/Projects/git-wntr/wntr/epanet/toolkit.py", line 156, in _error
wntr.epanet.toolkit.EpanetException: EPANET Error 305
>>> for i in range(0, 2000):
...     sim = wntr.sim.EpanetSimulator(network)
...     results = sim.run_sim(save_hyd=True)
...     print(i)
... 
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
  File "/Users/dbhart/Projects/git-wntr/wntr/sim/epanet.py", line 75, in run_sim
  File "/Users/dbhart/Projects/git-wntr/wntr/network/model.py", line 1760, in write_inpfile
  File "/Users/dbhart/Projects/git-wntr/wntr/epanet/io.py", line 422, in write
OSError: [Errno 24] Too many open files: 'temp.inp'
>>> 
```

**Comment by tasseff** [comment]
Thank you very much! I appreciate the alternatives you've suggested and will try the first in the meantime.

**Comment by kaklise** [comment]
This issues has been resolved with the addition of EPANET 2.2 in WNTR.
